python/tvm/topi/arm_cpu/mprofile/dsp/micro_kernel/tensordot.py

```diff
@@ -14,142 +14,333 @@
+            yield f"sum_{index} = __builtin_arm_smlatb({op2}, {op1}, sum_{index});"
+        else:
+            yield f"sum_{index} = __builtin_arm_{instruction}({op1}, {op2}, sum_{index});"
+
+
+def _requantize_sums(num_sums, requantize_shift, output_zero_point) -> Iterator[str]:
+    """Generates code to requantize the accumulator values.
+
+    The generated code does not use floating point instructions, as it simulates floating point
+    multiplication with an a int64 multiply + shift. The bias is added at the beginning, so we can
+    skip doing it now. The shift is hard-coded, as this saves a few cycles without hurting accuracy
+    in "most" cases.
+
+    It's *possible* we could save one more cycle here by pre-multiplying the bias with the
+    requantize multiplier, and then doing the bias addition and shift in the same cycle (via <op2>).
+    However, it's complicated and only saves one cycle.
+
+    It's also worth noting the SSAT16 operation doesn't help us here. The data isn't stored as two
+    halfwords in a word, and rearrainging it would take at least one cycle. Two SSAT operations is
+    just as good.
+
+    Calling __builtin_arm_ssat directly is a little bit gross, but GCC and Clang are unreliable
+    about compiling other ways of writing this. Both the multiply + shift and shift + saturation
+    combine to one instruction each.
+    """
+
+    yield "int scale_val = *scale;"
+    for i in range(num_sums):
+        yield f"int requant_{i} = (sum_{i} * (long long) scale_val) >> {requantize_shift - 1};"
+        yield f"requant_{i} = (requant_{i} + 1) >> 1;"
+        yield f"requant_{i} = __builtin_arm_ssat(requant_{i} + {output_zero_point}, 8);"
+
+
+def _write_sums_to_memory(num_sums, offset, stride) -> Iterator[str]:
+    """Generates code to write the requantized sums to memory.
+
+    Note - halfword packing here *does* help. It seems
+    like it wouldn't, as doing two pipelined int16 stores takes two cycles - the same as halfword
+    packing plus a pipelined int32 store. We still do the int16 stores when there is an output
+    stride, though.
+
+    However, this lets the compiler re-order instructions to better preserve memory, as it doesn't
+    like breaking apart the store instructions (as this messes up pipelining).
+    """
+
+    if stride > 1:
+        for i in range(num_sums):
+            yield f"((short*) output)[{i * stride + offset}] = (short) requant_{i};"
 
     else:
-        raise ValueError(f"No tensordot implementation exists for dtype '{in_dtype}'!")
+        num_halfwords = (num_sums - offset) // 2
+        for i in range(num_halfwords):
+            index = 2 * i + offset
+            yield f"int packed_res_{i} = requant_{index} + (requant_{index + 1} << 16);"
 
-    function_name = _get_func_name(in_dtype, tensor_h, jump, tensor_w, suffix)
-    return textwrap.dedent(
-        (
-            f"""
-        #include <stdint.h>
-        #include <arm_nnsupportfunctions.h>
```

Comment:
Yay! I think this solves the same problem as https://github.com/apache/tvm/pull/13363 😸 !